internal/controlplane/handlers_reconciliationtasks.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2024 Stacklok, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controlplane
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"github.com/ThreeDotsLabs/watermill/message"
+	"github.com/google/uuid"
+	"github.com/rs/zerolog/log"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/stacklok/minder/internal/engine"
+	"github.com/stacklok/minder/internal/logger"
+	"github.com/stacklok/minder/internal/reconcilers"
+	pb "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
+)
+
+// CreateEntityReconciliationTask creates a task to reconcile the state of an entity
+func (s *Server) CreateEntityReconciliationTask(ctx context.Context,
+	in *pb.CreateEntityReconciliationTaskRequest) (
+	*pb.CreateEntityReconciliationTaskResponse, error,
+) {
+	// Populated by EntityContextProjectInterceptor using incoming request
+	entityCtx := engine.EntityFromContext(ctx)
+	err := entityCtx.Validate(ctx, s.store)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "error in entity context: %v", err)
+	}
+
+	entity := in.GetEntity()
+
+	if entity == nil {
+		return nil, status.Error(codes.InvalidArgument, "entity is required")
+	}
+
+	var msg *message.Message
+	var topic string
+
+	// TODO: Support other entity types, replace with switch
+	if entity.GetType() == pb.Entity_ENTITY_REPOSITORIES {
+		msg, err = s.getRepositoryReconciliationMessage(ctx, entity)
+		if err != nil {
+			return nil, err
+		}
+		topic = reconcilers.InternalReconcilerEventTopic
```

Comment:
I think we need to check that `entity` is owned by `entityCtx`, or we get a repeat of https://github.com/stacklok/minder/security/advisories/GHSA-v627-69v2-xx37.  (Yeah, this is trickier than it should be...)

Reply:
We probably also want to check that `entity` exists here, so at least the query for ownership also lets us alert if the target doesn't exist.

Reply:
Oh, it turns out that `getRepositoryReconciliationMessage` actually uses the server's database handle to fetch the repository!

Can we pass the `store` in explicitly, and make `getRepositoryReconciliationMessage` a free function (not associated with the `Server` struct)?  Right now, there's a lot of functionality tied up in `Server`, and I'd like to push towards _less_ coupling.

Reply:
> we get a repeat of [GHSA-v627-69v2-xx37](https://github.com/stacklok/minder/security/advisories/GHSA-v627-69v2-xx37)

I also thought of that vulnerability with `GetRepositoryByID(ctx, repoUUID)`. But the only reason why I didn't check the owner was assuming that `id` is an internal state (sort of secretive), unlike the public repo name. If we want to verify that only the owner can fetch the repo using `id`, we should _add the project and provider to the query_ itself and not hack around.